File: rust/src/folding.rs
use crate::commitment::CommitmentScheme;
use crate::field::Field;
use crate::matrix::MatrixOps;
use crate::r1cs::{R1CSShape, RelaxedR1CSInstance, RelaxedR1CSWitness};
use sha2::{Digest, Sha256};

/// Compute cross-term T = Az1.Bz2 + Az2.Bz1 - u1*Cz2 - u2*Cz1.
pub fn compute_cross_term<F: Field, M: MatrixOps<F>, C: CommitmentScheme<F>>(
    shape: &R1CSShape<F, M>,
    inst1: &RelaxedR1CSInstance<F, C>,
    wit1: &RelaxedR1CSWitness<F>,
    inst2: &RelaxedR1CSInstance<F, C>,
    wit2: &RelaxedR1CSWitness<F>,
) -> Vec<F> {
    let z1 = shape.make_relaxed_z(inst1.u, &inst1.x, &wit1.w);
    let z2 = shape.make_relaxed_z(inst2.u, &inst2.x, &wit2.w);

    let az1 = shape.a.mul_vec(&z1);
    let bz1 = shape.b.mul_vec(&z1);
    let az2 = shape.a.mul_vec(&z2);
    let bz2 = shape.b.mul_vec(&z2);
    let cz1 = shape.c.mul_vec(&z1);
    let cz2 = shape.c.mul_vec(&z2);

    (0..shape.m)
        .map(|i| {
            az1[i]
                .mul(bz2[i])
                .add(az2[i].mul(bz1[i]))
                .sub(inst1.u.mul(cz2[i]))
                .sub(inst2.u.mul(cz1[i]))
        })
        .collect()
}

/// Derive Fiat-Shamir challenge from transcript.
///
/// Transcript: com_T || x1 || x2 || u1 (8 bytes BE) || u2 (8 bytes BE) || com_W1 || com_W2
pub fn fiat_shamir_challenge<F: Field, C: CommitmentScheme<F>>(
    com_t: &C,
    inst1: &RelaxedR1CSInstance<F, C>,
    inst2: &RelaxedR1CSInstance<F, C>,
) -> F {
    let mut hasher = Sha256::new();
    hasher.update(com_t.value());
    for elem in &inst1.x {
        hasher.update(elem.to_bytes());
    }
    for elem in &inst2.x {
        hasher.update(elem.to_bytes());
    }
    hasher.update(inst1.u.to_bytes());
    hasher.update(inst2.u.to_bytes());
    hasher.update(inst1.com_w.value());
    hasher.update(inst2.com_w.value());

    let hash = hasher.finalize();
    let r_bytes: [u8; 8] = hash[..8].try_into().unwrap();
    let r_int = u64::from_be_bytes(r_bytes);
    F::from_u64(r_int)
}

/// Fold two Relaxed R1CS instances into one.
pub fn fold<F: Field, M: MatrixOps<F>, C: CommitmentScheme<F>>(
    shape: &R1CSShape<F, M>,
    inst1: &RelaxedR1CSInstance<F, C>,
    wit1: &RelaxedR1CSWitness<F>,
    inst2: &RelaxedR1CSInstance<F, C>,
    wit2: &RelaxedR1CSWitness<F>,
) -> (RelaxedR1CSInstance<F, C>, RelaxedR1CSWitness<F>) {
    // Step 1: Cross-term
    let t = compute_cross_term(shape, inst1, wit1, inst2, wit2);

    // Step 2: Commit to T
    let com_t = C::commit(&t);

    // Step 3: Challenge
    let r = fiat_shamir_challenge(&com_t, inst1, inst2);
    let r_sq = r.mul(r);

    // Step 4: Fold
    // E' = E1 + r*T + r^2*E2
    let e_folded: Vec<F> = (0..shape.m)
        .map(|i| wit1.e[i].add(r.mul(t[i])).add(r_sq.mul(wit2.e[i])))
        .collect();

    // u' = u1 + r*u2
    let u_folded = inst1.u.add(r.mul(inst2.u));

    // x' = x1 + r*x2
    let x_folded: Vec<F> = inst1
        .x
        .iter()
        .zip(inst2.x.iter())
        .map(|(&a, &b)| a.add(r.mul(b)))
        .collect();

    // W' = W1 + r*W2
    let w_folded: Vec<F> = wit1
        .w
        .iter()
        .zip(wit2.w.iter())
        .map(|(&a, &b)| a.add(r.mul(b)))
        .collect();

    let com_e_folded = C::commit(&e_folded);
    let com_w_folded = C::commit(&w_folded);

    let folded_instance = RelaxedR1CSInstance {
        com_e: com_e_folded,
        u: u_folded,
        x: x_folded,
        com_w: com_w_folded,
    };
    let folded_witness = RelaxedR1CSWitness {
        e: e_folded,
        w: w_folded,
    };

    (folded_instance, folded_witness)
}

File: nano_nova/folding.py
"""
Nova's NIFS (Non-Interactive Folding Scheme) for Relaxed R1CS.

This is the heart of Nova. Given two Relaxed R1CS instances, we fold them
into a single instance of the same size. The cross-term T absorbs the
"interference" from combining two quadratic constraints.

The physics analogy: T is a renormalization counter-term that absorbs
perturbative corrections when combining two constraint systems.

Nova paper reference: Section 4, Construction 4 (NIFS for Relaxed R1CS)
"""

import hashlib

import numpy as np

from .commitment import ToyCommitment, commit
from .field import GF
from .r1cs import R1CSShape, RelaxedR1CSInstance, RelaxedR1CSWitness


def compute_cross_term(
    shape: R1CSShape,
    instance1: RelaxedR1CSInstance,
    witness1: RelaxedR1CSWitness,
    instance2: RelaxedR1CSInstance,
    witness2: RelaxedR1CSWitness,
) -> np.ndarray:
    """Compute the cross-term T for folding two relaxed instances.

    T = Az₁ ∘ Bz₂ + Az₂ ∘ Bz₁ - u₁·Cz₂ - u₂·Cz₁

    This is the "interference" term that arises from the quadratic nature
    of R1CS when combining two instances via random linear combination.

    Nova paper: Equation in Construction 4, step 1.

    Args:
        shape: R1CS structure (A, B, C matrices).
        instance1, witness1: First relaxed instance.
        instance2, witness2: Second relaxed instance.

    Returns:
        Cross-term vector T of length m (number of constraints).
    """
    z1 = shape._make_relaxed_z(instance1.u, instance1.x, witness1.W)
    z2 = shape._make_relaxed_z(instance2.u, instance2.x, witness2.W)

    Az1 = shape.A @ z1
    Bz1 = shape.B @ z1
    Az2 = shape.A @ z2
    Bz2 = shape.B @ z2
    Cz1 = shape.C @ z1
    Cz2 = shape.C @ z2

    # T = Az₁ ∘ Bz₂ + Az₂ ∘ Bz₁ - u₁·Cz₂ - u₂·Cz₁
    T = Az1 * Bz2 + Az2 * Bz1 - instance1.u * Cz2 - instance2.u * Cz1
    return T


def fiat_shamir_challenge(
    com_T: ToyCommitment,
    instance1: RelaxedR1CSInstance,
    instance2: RelaxedR1CSInstance,
) -> np.ndarray:
    """Derive a challenge r via Fiat-Shamir (hash of transcript).

    In the interactive protocol, the verifier sends random r after seeing com_T.
    Fiat-Shamir makes this non-interactive by hashing the transcript.

    Note: the full Nova transcript (Construction 4) also includes com_E1 and
    com_E2. We omit them here for simplicity — this is sufficient for the
    educational IVC demo but not for production security.

    Nova paper: Section 4, making NIFS non-interactive.

    Args:
        com_T: Commitment to the cross-term.
        instance1: First public instance.
        instance2: Second public instance.

    Returns:
        Challenge r as a GF(p) field element.
    """
    # 9 bytes covers the full 61-bit field prime (2^61 - 1 < 2^72 = 9 bytes)
    transcript = b""
    transcript += com_T.value
    transcript += instance1.x.tobytes()
    transcript += instance2.x.tobytes()
    transcript += int(instance1.u).to_bytes(9, "big")
    transcript += int(instance2.u).to_bytes(9, "big")
    transcript += instance1.com_W.value
    transcript += instance2.com_W.value

    h = hashlib.sha256(transcript).digest()
    r_int = int.from_bytes(h[:8], "big") % int(GF.order)
    return GF(r_int)


def fold(
    shape: R1CSShape,
    instance1: RelaxedR1CSInstance,
    witness1: RelaxedR1CSWitness,
    instance2: RelaxedR1CSInstance,
    witness2: RelaxedR1CSWitness,
) -> tuple[RelaxedR1CSInstance, RelaxedR1CSWitness]:
    """Fold two Relaxed R1CS instances into one.

    This is the core Nova operation. Given (inst₁, w₁) and (inst₂, w₂):
    1. Compute cross-term T
    2. Commit to T
    3. Derive challenge r (Fiat-Shamir)
    4. Fold: E' = E₁ + r·T + r²·E₂
             u' = u₁ + r·u₂
             x' = x₁ + r·x₂
             W' = W₁ + r·W₂

    The folded instance satisfies Relaxed R1CS if both inputs did.

    Nova paper: Construction 4 (NIFS.P and NIFS.V).

    Args:
        shape: R1CS structure.
        instance1, witness1: First relaxed instance (typically the running accumulator).
        instance2, witness2: Second relaxed instance (typically a fresh step).

    Returns:
        (folded_instance, folded_witness) — a new valid Relaxed R1CS pair.
    """
    # Step 1: Compute cross-term
    T = compute_cross_term(shape, instance1, witness1, instance2, witness2)

    # Step 2: Commit to T
    com_T = commit(T)

    # Step 3: Fiat-Shamir challenge
    r = fiat_shamir_challenge(com_T, instance1, instance2)

    # Step 4: Fold everything
    # E' = E₁ + r·T + r²·E₂
    r_sq = r * r
    E_folded = witness1.E + r * T + r_sq * witness2.E

    # u' = u₁ + r·u₂
    u_folded = instance1.u + r * instance2.u

    # x' = x₁ + r·x₂
    x_folded = instance1.x + r * instance2.x

    # W' = W₁ + r·W₂
    W_folded = witness1.W + r * witness2.W

    # Commitments fold homomorphically (simulated here)
    com_E_folded = commit(E_folded)
    com_W_folded = commit(W_folded)

    folded_instance = RelaxedR1CSInstance(
        com_E=com_E_folded,
        u=u_folded,
        x=x_folded,
        com_W=com_W_folded,
    )
    folded_witness = RelaxedR1CSWitness(E=E_folded, W=W_folded)

    return folded_instance, folded_witness
